### ISSUE TITLE
FR: Better dependencies re: test vs non-test

### ISSUE THREAD

**Issue by virusdave** [zio/zio-prelude#604]
[redirecting from https://github.com/zio/zio/issues/4952]

# Requested: 
`zio-prelude` as a non-test library does not have direct compile-time dependencies on "testonly" libraries like `zio-test`.

# Background:
zio-prelude has a compile (non-test) dependency on zio-test.  This makes usage in Bazel world more painful, since bazel differentiates "testonly" targets from "other" targets, and prevents any "other" target from depending on any "testonly" target.  This is good, and helps prevent another [Knight Capital situation](https://www.cio.com/article/2393212/software-testing-lessons-learned-from-knight-capital-fiasco.html) from happening.

It looks like zio-prelude has a dependency on zio-test due to (among other things) generators, or possibly other functionality that lives in zio-test that is useful outside of a testing context.

A couple of approaches are possible:
1 Split `zio-test` functionality into truly "testonly" and "non-test" functionality libraries, then depend (as appropriate) on those distinctly.
2 If `zio-prelude` is actually using the `zio-test` functionality in the context of tests or testing-related libraries (ie: non-prod), split `zio-prelude`'s use and either make `zio-test` a testing dependency, or possibly split the testing functionality into a `zio-prelude-test` library and move the `zio-test` dependency there?

There may be other approaches that are reasonable; i'm definitely not an expert on what `zio-prelude` is doing with these dependencies.

## Conversation
See [this conversation](https://discord.com/channels/629491597070827530/630498701860929559/831197477633523733)

**Comment by adamgfraser** [comment]
The dependency is basically used in two places:

1. Assertions are used to support validation of data values
2. Laws and generators are used for defining the behavior of various abstractions

For the first we are already in the process of moving to separate macro powered assertions. For the second I think we could potentially move those laws into a separate hierarchy though it is not as nice in some ways as each abstraction no longer comes with its own executable set of laws.

**Comment by virusdave** [comment]
For 2, would making the dependency a ["test scope"](https://www.baeldung.com/maven-dependency-scopes#4-test) dependency be an option?

**Comment by adamgfraser** [comment]
I'm not sure what the semantics of that are. The code is in places like [here](https://github.com/zio/zio-prelude/blob/5dd8c655dcf2e8cb07c8073eb7d3552c0403d9fc/core/shared/src/main/scala/zio/prelude/Associative.scala#L90). So there are lots of methods on this object that you definitely might want to call in production code in addition to these laws. What happens when you call one of these methods. Does that still compile and run correctly?

**Comment by virusdave** [comment]
Ah, interesting, so probably not?  I'm definitely not an expert on Prelude's design, so maybe i'm suggesting the wrong locations for consideration of cuts or refactoring.  It might then be the upstream components that should be split?  Perhaps Laws are non test?  Or Generators are non test?  But specific uses of those ARE testing-only?

Given what you know of Prelude's design, is it even possible to separate notions into "this is real / for production use" and "this is for testing the correctness of the former, but should not be included as part of production use" ?

**Comment by adamgfraser** [comment]
I think conceptually it basically comes down to what you think of this design:

```scala
trait TypeClass {
  // various methods
}

object TypeClass {

  val laws = ???

  // various constructors and instances
}
```

The benefit of this is that a type class essentially "comes with" a way of testing that instances satisfy its laws, which has some conceptual appeal. But it does mix "production" and "test" code to some extend versus a more traditional encoding such as:

```scala
// in one module
package zio.prelude

trait TypeClass {
  // various methods
}

object TypeClass {

  // various constructors and instances
}

// in some other "test only" module

package zio.prelude.test

object TypeClass {
  val laws = ???
}
```

**Comment by sideeffffect** [comment]
Wouldn't the simplest solution (at least short- to mid-term) be to treat `zio-test` as a production artifact, not a mere test artifact?

Moving laws from `Compile` to `Test` configuration is something we could at least consider.

**Comment by adamgfraser** [comment]
I think the problem is that at least in the Bazel world you want to enforce this very strong separation between test and production dependencies so you want ZIO Test to be a test only dependency so that anything else that depends on ZIO Test can't be included in your production code so you can't accidentally use test code in production.

**Comment by sideeffffect** [comment]
Yes, so what is stopping Bazel users from treating, temporarily, ZIO Test as a production dependency?

**Comment by adamgfraser** [comment]
I don't think anything is stopping them. My understanding is that is the workaround they are using today. They just don't like it. 😃

**Comment by sideeffffect** [comment]
:smile_cat: Oh, OK! So we have some time to figure out if we want to move laws out, right?
I guess we could do that, cats do that too, if I'm not mistaken.

**Comment by adamgfraser** [comment]
Yes I think that is right.

**Comment by virusdave** [comment]
> I think conceptually it basically comes down to what you think of this design:
> 
> ```scala
> trait TypeClass {
>   // various methods
> }
> 
> object TypeClass {
> 
>   val laws = ???
> 
>   // various constructors and instances
> }
> ```

This seems sensible to me, assuming that `laws` refers to interfaces or implementations that are sensibly defined in non-test libraries and aren't intrinsically risky to include in non-test deployed code.  I'd have to look more into the details of `Laws` to speak more intelligently about it (which i'll do when I have some free time, eventually...).  If tests that assert that the laws aren't being violated are able to be segmented away from the laws themselves, this seems like it might be just fine.

For instance, if a law like `ForAll[T], F[T].map(id) === F[T]` can be expressed in non-test world, but the _implementation_ of `ForAll[T]` that actually executes during testing time is constrained to a testing-only library, that should work.

I might be talking out of my ass until i learn more :)  I appreciate your consideration though!

**Comment by adamgfraser** [comment]
Closed by #743.
